### PR TITLE
Remote Validation

### PR DESCRIPTION
**Pull request by hendrikebbers**
First classes of the remote validation module

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/838)

### CODE REVIEW

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+package com.canoo.dp.impl.validation.rest;
+
+/**
+ * @author Gunnar Morling
+ * @author Hendrik Ebbers
+ */
+public class InvalidJsonException extends Exception {
+
+    public InvalidJsonException(String message) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        super(message);
+    }
+
+    public InvalidJsonException(String message, Throwable cause) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    public RemoteValidationServlet(final RemoteValidationConfiguration configuration) {
+        this(configuration, Validation.buildDefaultValidatorFactory().getValidator());
+    }
+
+    public RemoteValidationServlet(final RemoteValidationConfiguration configuration, final Validator validator) {
+        this(configuration, validator, RemoteValidationServlet.class.getClassLoader());
+    }
+
+    public RemoteValidationServlet(final RemoteValidationConfiguration configuration, final Validator validator, final ClassLoader classLoader) {
+        this.remoteValidator = new RemoteValidator(configuration, validator, classLoader);
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java]: Should we add some not-null checks into this constructor?

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    private Set<ConstraintViolation<?>> validate(final String typeName, final String propertyName, final JsonElement propertyElement) {
+        Assert.requireNonNull(propertyElement, "propertyElement");
+        if (!propertyElement.isJsonPrimitive() && !propertyElement.isJsonNull()) {
+            throw new IllegalArgumentException("Only primitive or null properties supported. Wrong type for " + typeName + "." + propertyName);
+        }
+        if (propertyElement.isJsonNull()) {
+            return remoteValidator.validateValue(typeName, propertyName, null);
+        }
+        JsonPrimitive primitiveValue = propertyElement.getAsJsonPrimitive();
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    private final Map<Class<? extends Annotation>, String> constraintAnnotationMapping = new HashMap<>();
+
+    private final Map<String, Class<?>> typeMapping = new HashMap<>();
+
+    public DefaultValidationConfiguration() {
+    }
+
+    //TODO: Rename
+    public void setAddViolationMessageToResponse(boolean addViolationMessageToResponse) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        this.addViolationMessageToResponse = addViolationMessageToResponse;
+    }
+
+    //TODO: Rename
+    public void setAddViolationMessageTemplateToResponse(boolean addViolationMessageTemplateToResponse) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    private boolean addViolationMessageTemplateToResponse = false;
+
+    private final Map<Class<? extends Annotation>, String> constraintAnnotationMapping = new HashMap<>();
+
+    private final Map<String, Class<?>> typeMapping = new HashMap<>();
+
+    public DefaultValidationConfiguration() {
+    }
+
+    //TODO: Rename
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java]: WIP ? Remove TODO and change the name if possible.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    public void setAddViolationMessageToResponse(boolean addViolationMessageToResponse) {
+        this.addViolationMessageToResponse = addViolationMessageToResponse;
+    }
+
+    //TODO: Rename
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java]: WIP ? Remove TODO and change the name if possible.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    public Set<ConstraintViolation<?>> validateValue(final String typeName, final String property, final Object value) {
+        Assert.requireNonNull(typeName, "typeName");
+        Assert.requireNonNull(property, "property");
+        return validator.validateValue((Class)getClassForType(typeName), property, value);
+    }
+
+    private Class<?> getClassForType(final String typeName) {
+        Assert.requireNonNull(typeName, "typeName");
+        Map<String, Class<?>> mapping = configuration.getTypeMapping();
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        if(mapping.containsKey(typeName)) {
+            Class<?> cls = mapping.get(typeName);
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+            if(cls == null) {
+                throw new IllegalStateException("null value specified for type " + typeName);
+            }
+            return cls;
+        }
+        try {
+            return classLoader.loadClass(typeName);
+        } catch (ClassNotFoundException e) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java (hunk 1 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java]: Done.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java (hunk 2 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/InvalidJsonException.java]: Done.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java (hunk 3 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java]: Done.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java (hunk 4 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/RemoteValidationServlet.java]: Done.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java (hunk 6 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java]: Done.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java (hunk 8 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/DefaultValidationConfiguration.java]: @hendrikebbers what should be the name?

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java (hunk 9 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java]: Done.

File: platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java (hunk 11 above)
Comment (kunsingh) [reply platform-extras/dolphin-platform-bean-validation-rest/src/main/java/com/canoo/dp/impl/validation/rest/validator/impl/RemoteValidator.java]: Done.

**Pull request merged**